src/main/java/io/reactivex/internal/subscribers/BoundedSubscriber.java

```diff
@@ -69,12 +67,14 @@ public void onSubscribe(Subscription s) {
     public void onNext(T t) {
         if (!isDisposed()) {
             try {
-                if (internalBuffer.getAndIncrement() < bufferSize) {
-                    onNext.accept(t);
-                }
-                if (internalBuffer.get() == bufferSize) {
-                    onComplete.run();
-                    get().cancel();
+                onNext.accept(t);
+
+                int c = consumed + 1;
+                if (c == bufferSize) {
```

Comment:
It is better to use a low-watermark than requesting only when the entire buffer has been consumed. Define a `final int limit` field, in the constructor set it as `this.limit = bufferSize - (bufferSize >> 2);`, then compare and request as `limit`.